crates/matrix-sdk-base/src/response_processors/state_events.rs

```diff
@@ -248,16 +307,762 @@ where
+                    event_factory.create(sender, RoomVersionId::try_from("45").unwrap()),
+                ))
+                .add_joined_room(JoinedRoomBuilder::new(room_id_1).add_timeline_event(
+                    event_factory.create(sender, RoomVersionId::try_from("46").unwrap()),
+                ))
+                .add_joined_room(JoinedRoomBuilder::new(room_id_2).add_timeline_event(
+                    event_factory.create(sender, RoomVersionId::try_from("47").unwrap()),
+                ))
+                .build_sync_response();
+
+            assert!(client.receive_sync_response(response).await.is_ok());
+
+            // Room 0
+            // the third `m.room.create` has been ignored!
+            assert_eq!(
+                client.get_room(room_id_0).unwrap().create_content().unwrap().room_version.as_str(),
+                "42"
+            );
+            // Room 1
+            // the second `m.room.create` has been ignored!
+            assert_eq!(
+                client.get_room(room_id_1).unwrap().create_content().unwrap().room_version.as_str(),
+                "44"
+            );
+            // Room 2
+            assert_eq!(
+                client.get_room(room_id_2).unwrap().create_content().unwrap().room_version.as_str(),
+                "47"
+            );
+        }
+    }
+
+    #[async_test]
+    async fn test_check_room_upgrades_no_newly_tombstoned_rooms() {
+        let client = logged_in_base_client(None).await;
+
+        // Create a new room, no tombstone, no anything.
+        {
+            let response = SyncResponseBuilder::new()
+                .add_joined_room(JoinedRoomBuilder::new(room_id!("!r0")))
+                .build_sync_response();
+
+            assert!(client.receive_sync_response(response).await.is_ok());
+        }
+    }
+
+    #[async_test]
+    async fn test_check_room_upgrades_no_error() {
+        let sender = user_id!("@mnt_io:matrix.org");
+        let event_factory = EventFactory::new().sender(sender);
+        let mut response_builder = SyncResponseBuilder::new();
+        let room_id_0 = room_id!("!r0");
+        let room_id_1 = room_id!("!r1");
+        let room_id_2 = room_id!("!r2");
+
+        let client = logged_in_base_client(None).await;
+
+        // Room 0.
+        {
+            let response = response_builder
+                .add_joined_room(JoinedRoomBuilder::new(room_id_0).add_timeline_event(
+                    // Room 0 has no predecessor.
+                    event_factory.create(sender, RoomVersionId::try_from("41").unwrap()),
+                ))
+                .build_sync_response();
+
+            assert!(client.receive_sync_response(response).await.is_ok());
+
+            let room_0 = client.get_room(room_id_0).unwrap();
+
+            assert!(room_0.predecessor_room().is_none());
+            assert!(room_0.successor_room().is_none());
+        }
+
+        // Room 0 <-> room 1.
```

Comment:
The `<->` confused me here. I understand why it is double-headed, but it reads like it's a loop.

Could do: `Room 0 --tombstone-->    <--create-- Room 1` ?